core/os/thread.cpp

```diff
@@ -71,31 +67,23 @@ void Thread::callback(ID p_caller_id, const Settings &p_settings, Callback p_cal
 }
 
 void Thread::start(Thread::Callback p_callback, void *p_user, const Settings &p_settings) {
-	if (id != _thread_id_hash(std::thread::id())) {
-#ifdef DEBUG_ENABLED
-		WARN_PRINT("A Thread object has been re-started without wait_to_finish() having been called on it. Please do so to ensure correct cleanup of the thread.");
-#endif
-		thread.detach();
-		std::thread empty_thread;
-		thread.swap(empty_thread);
-	}
-	std::thread new_thread(&Thread::callback, _thread_id_hash(thread.get_id()), p_settings, p_callback, p_user);
+	ERR_FAIL_COND_MSG(id != UNASSIGNED_ID, "A Thread object has been re-started without wait_to_finish() having been called on it.");
+	id = id_counter.increment();
+	std::thread new_thread(&Thread::callback, id, p_settings, p_callback, p_user);
 	thread.swap(new_thread);
-	id = _thread_id_hash(thread.get_id());
 }
 
 bool Thread::is_started() const {
-	return id != _thread_id_hash(std::thread::id());
+	return id != UNASSIGNED_ID;
 }
 
 void Thread::wait_to_finish() {
-	if (id != _thread_id_hash(std::thread::id())) {
-		ERR_FAIL_COND_MSG(id == get_caller_id(), "A Thread can't wait for itself to finish.");
-		thread.join();
-		std::thread empty_thread;
-		thread.swap(empty_thread);
-		id = _thread_id_hash(std::thread::id());
-	}
+	ERR_FAIL_COND_MSG(id == UNASSIGNED_ID, "Attempt of waiting to finish on a thread that was never started.");
```

Comment:
I guess if it errors its because its broken and its now being detected. It should not crash though, that is weird.

Reply:
It doesn't crash, I changed it to a CRASH_COND just to get the full backtrace.

Reply:
It should not be broken. But both cases should have a check if the thread is running.

`RemoteDebuggerPeerTCP` calls `wait_to_finish` even if it is not connected (and thread is not started).
`EditorExportPlatformAndroid` probably calls it when a temporary instance created by editor help (to get default values) is destroyed.

After a quick look, there are some other places that might have the similar behavior as well.

Reply:
The problem here is that the bespoke thread id is lazily generated in `get_caller_id()`. Threads on which it hasn't been called will error here and also have false negatives in the check below. My suggestion is that the bespoke id is generated on `start()`.

Reply:
For a foregin thread (attached), it's OK to generate the id lazily, but a flag should be set (a `bool` or just reserve the MSB of the id) to disable the regular wait and destructor checks (and error if it's tried to be awaited).

Reply:
@RandomShaper There is no way to wait on those threads or them destructing because there is no thread object, so I think it should be fine.

Reply:
@bruvzg I think those classes should not wait if the thread is not active, IMO it seems more reasonable to fix those cases by checking whether the thread has started than bypassing that check. I can do this I guess.

Reply:
Oh, right. I thought there was a way to obtain a thread object from a foreign thread. It turns out I made it all up. 😝

Reply:
But this is still relevant;

> The problem here is that the bespoke thread id is lazily generated in `get_caller_id()`. Threads on which it hasn't been called will error here and also have false negatives in the check below. My suggestion is that the bespoke id is generated on `start()`.

Reply:
I am resolving this conversation, I fixed the errors that were reported because I think this should be effectively reported as an error. If more are reported due to this they will have to be fixed too in their respective places.